providers/dynadot/api.go

```diff
@@ -0,0 +1,135 @@
+package dynadot
+
+import (
+	"encoding/xml"
+	"fmt"
+	"io"
+	"net/http"
+	"strings"
+)
+
+// API layer for Dynadot
+
+type dynadotProvider struct {
+	key string
+}
+
+type requestParams map[string]string
+
+type header struct {
+	SuccessCode int    `xml:"SuccessCode"`
+	Status      string `xml:"Status"`
+	Error       string `xml:"Error,omitempty"`
+}
+
+type addNsResponse struct {
+	XMLName     xml.Name `xml:"AddNsResponse"`
+	AddNsHeader header   `xml:"AddNsHeader"`
+}
+
+type setNsResponse struct {
+	XMLName     xml.Name `xml:"SetNsResponse"`
+	SetNsHeader header   `xml:"SetNsHeader"`
+}
+
+type getNsResponse struct {
+	XMLName     xml.Name  `xml:"GetNsResponse"`
+	NsContent   nsContent `xml:"NsContent"`
+	GetNsHeader header    `xml:"GetNsHeader"`
+}
+
+type nsContent struct {
+	Host   []string `xml:"Host"`
+	NsName string   `xml:"NsName"`
+}
+
+func (c *dynadotProvider) getNameservers(domain string) ([]string, error) {
+	var bodyString, err = c.get("get_ns", requestParams{"domain": domain})
+	if err != nil {
+		return []string{}, fmt.Errorf("failed NS list (Dynadot): %s", err)
+	}
+	var ns getNsResponse
+	xml.Unmarshal(bodyString, &ns)
+
+	if ns.GetNsHeader.SuccessCode != 0 {
+		return []string{}, fmt.Errorf("failed NS list (Dynadot): %s", ns.GetNsHeader.Error)
+	}
+
+	hosts := []string{}
+	hosts = append(hosts, ns.NsContent.Host...)
+	return hosts, nil
+}
+
+func (c *dynadotProvider) updateNameservers(ns []string, domain string) error {
+	if len(ns) > 13 {
+		return fmt.Errorf("failed NS update (Dynadot): only up to 13 nameservers are supported")
+	}
+
+	// Nameservers must first be added to the Dynadot account
+	for _, host := range ns {
+		b, err := c.get("add_ns", requestParams{"host": host})
+		if err != nil {
+			return fmt.Errorf("failed NS add (Dynadot): %s", err)
+		}
+		var resp addNsResponse
+		err = xml.Unmarshal(b, &resp)
+		if err != nil {
+			return fmt.Errorf("failed NS add (Dynadot): %s", err)
+		}
+
+		if resp.AddNsHeader.SuccessCode != 0 {
+			// No apparent way to get all existing entries on an account, so filter
+			if strings.Contains(resp.AddNsHeader.Error, "already exists") {
+				continue
+			}
```

Comment:
The Dynadot API requires nameservers be added to a separate list on each account before they can be added to a domain.

I don't see any way with their API to get this full list, so I've just attempted to add every nameserver to it every time, ignoring ones that error due to already existing.

This feels super hacky although I don't see a better way of doing it.

The SuccessCode is always just -1 for a failure, there are no codes for specific cases so checking the error message is required.

Reply:
I can't think of anything better.